### ISSUE TITLE
About Resblock x / math.sqrt(2)

### ISSUE THREAD

**Issue by zhangqianhui** [clovaai/stargan-v2#6]
I am confused that the output of ResBlk functions would divide 'math.sqrt(2)'
@yunjey @youngjung

**Comment by youngjung** [comment]
Var(`out + shortcut` / sqrt(2)) = 1/2*(Var(`out`) + Var(`shortcut`)).
It makes the first steps of training more stable.
You can find a relationship to [He initialization](https://arxiv.org/abs/1502.01852).

**Comment by yunjey** [comment]
@zhangqianhui 

Let X and Y be independent random variables and Var(X)=Var(Y)=1, which is a common assumption in frequently used initialization techniques (e.g. He init, Xavier init). Then, the variance of the sum of the two variables Var(X+Y) = Var(X) + Var(Y) = 2, where the sum operation **doubles** the input variance.

This issue is not noticeable in classification tasks, since classification models use batch normalization which makes the variance of activation values to 1. On the other hand, in generation tasks, we use instance normalization, and thus the above issue could make the initial training very unstable. 

To address this issue, we can simply divide X + Y by sqrt(2) so that the variance Var((X+Y) / sqrt(2) = (Var(X) + Var(Y)) / 2 = 1. This kind of scaling trick has been used [here](https://github.com/LMescheder/GAN_stability/blob/master/gan_training/models/resnet.py#L133), [here](https://arxiv.org/abs/1901.09321), and [here](https://arxiv.org/abs/1912.04958).

**Comment by zhangqianhui** [comment]
Thanks. 

Additionally, in your paper, you have claimed that the original gan with non-saturating adversarial loss type. 

My understanding:

the gan loss for g: -log(D(G(x))) instead of log((1 - D(G(x)))).

In your code, you use the BCE loss:

> 
> targets = torch.full_like(logits, fill_value=target)
>     loss = F.binary_cross_entropy_with_logits(logits, targets)

By your loss, we would have: For brevity, let x = logits, z = labels. The logistic loss is

 z * -log(sigmoid(x)) + (1 - z) * -log(1 - sigmoid(x))

when z = 0, it will be:

-log(1 - sigmoid(x)) = -log(1 - D(G(x))).

It is not a non-saturating adversarial loss. 

I think using softplus is right in Spectral normalization GANs.

@yunjey @youngjung

**Comment by yunjey** [comment]
@zhangqianhui When training the generator, we use 1 as the target value not 0.

**Comment by zhangqianhui** [comment]
I am sorry. I misunderstand.

**Comment by zhangqianhui** [comment]
Hello, Why do you high-pass module for generator. I found that it is not explained in the original paper.

@yunjey  @youngjung

**Comment by yunjey** [comment]
@zhangqianhui The high-pass filter emphasizes boundary pixels between the contrast areas and is often called an edge detector. This is optional in a skip-connection. Similar results can be obtained without high-pass filtering.

**Comment by zhangqianhui** [comment]
Thanks!

**Comment by zhangqianhui** [comment]
Hello, Does the generator need tanh? I can not find it in this project.
@yunjey @youngjung

**Comment by jaejun-yoo** [comment]
@zhangqianhui Hello. Could you please open the question in a separate issue? I do not think this is a good way to maintain the thread that you reopen with different kinds of small questions that are not related to the original topic.
